File: src/main/java/kr/hahaha98757/zombiesaddon/data/GameMode.java
//Code in Zombies Utils by Stachelbeere1248

package kr.hahaha98757.zombiesaddon.data;

import kr.hahaha98757.zombiesaddon.enums.Difficulty;
import kr.hahaha98757.zombiesaddon.enums.Map;
import org.jetbrains.annotations.Contract;

public enum GameMode {
    DEAD_END(Map.DEAD_END, Difficulty.NORMAL), DEAD_END_HARD(Map.DEAD_END, Difficulty.HARD), DEAD_END_RIP(Map.DEAD_END, Difficulty.RIP),
    BAD_BLOOD(Map.BAD_BLOOD, Difficulty.NORMAL), BAD_BLOOD_HARD(Map.BAD_BLOOD, Difficulty.HARD), BAD_BLOOD_RIP(Map.BAD_BLOOD, Difficulty.RIP),
    ALIEN_ARCADIUM(Map.ALIEN_ARCADIUM, Difficulty.NORMAL),
    PRISON(Map.PRISON, Difficulty.NORMAL), PRISON_HARD(Map.PRISON, Difficulty.HARD), PRISON_RIP(Map.PRISON, Difficulty.RIP);

    private final Map map;
    private final Difficulty difficulty;

    GameMode(Map map, Difficulty difficulty) {
        this.map = map;
        this.difficulty = difficulty;
    }

    public Map getMap() {
        return map;
    }

    public Difficulty getDifficulty() {
        return difficulty;
    }

    public GameMode appliedDifficulty(Difficulty difficulty) {
        switch (map) {
            case DEAD_END:
                switch (difficulty) {
                    case NORMAL:
                        return DEAD_END;
                    case HARD:
                        return DEAD_END_HARD;
                    case RIP:
                        return DEAD_END_RIP;
                }
            case BAD_BLOOD:
                switch (difficulty) {
                    case NORMAL:
                        return BAD_BLOOD;
                    case HARD:
                        return BAD_BLOOD_HARD;
                    case RIP:
                        return BAD_BLOOD_RIP;
                }
            case PRISON:
                switch (difficulty) {
                    case NORMAL:
                        return PRISON;
                    case HARD:
                        return PRISON_HARD;
                    case RIP:
                        return PRISON_RIP;
                }
            case ALIEN_ARCADIUM:
                return ALIEN_ARCADIUM;
            default:
                return null;
        }
    }
    public boolean isMap(Map map) {
        return getMap() == map;
    }

    @Contract(pure = true)
    public static GameMode getNormalForMap(Map map) {
        switch (map) {
            case DEAD_END:
                return DEAD_END;
            case BAD_BLOOD:
                return BAD_BLOOD;
            case ALIEN_ARCADIUM:
                return ALIEN_ARCADIUM;
            case PRISON:
                return PRISON;
            default:
                return null;
        }
    }
}

File: src/main/kotlin/kr/hahaha98757/zombiesaddon/enums/GameMode.kt
package kr.hahaha98757.zombiesaddon.enums

import kr.hahaha98757.zombiesaddon.data.Round
import kr.hahaha98757.zombiesaddon.utils.JsonLoader

enum class GameMode(val map: ZombiesMap, val difficulty: Difficulty, val rounds: Array<Round>) {
    DEAD_END_NORMAL(ZombiesMap.DEAD_END, Difficulty.NORMAL, readFromFile("data/rounds/DEN_WAVES.json")),
    DEAD_END_HARD(ZombiesMap.DEAD_END, Difficulty.HARD, readFromFile("data/rounds/DEH_WAVES.json")),
    DEAD_END_RIP(ZombiesMap.DEAD_END, Difficulty.RIP, readFromFile("data/rounds/DER_WAVES.json")),

    BAD_BLOOD_NORMAL(ZombiesMap.BAD_BLOOD, Difficulty.NORMAL, readFromFile("data/rounds/BBN_WAVES.json")),
    BAD_BLOOD_HARD(ZombiesMap.BAD_BLOOD, Difficulty.HARD, readFromFile("data/rounds/BBH_WAVES.json")),
    BAD_BLOOD_RIP(ZombiesMap.BAD_BLOOD, Difficulty.RIP, readFromFile("data/rounds/BBR_WAVES.json")),

    ALIEN_ARCADIUM(ZombiesMap.ALIEN_ARCADIUM, Difficulty.NORMAL, readFromFile("data/rounds/AA_WAVES.json")),

    PRISON_NORMAL(ZombiesMap.PRISON, Difficulty.NORMAL, readFromFile("data/rounds/PRN_WAVES.json")),
    PRISON_HARD(ZombiesMap.PRISON, Difficulty.HARD, readFromFile("data/rounds/PRH_WAVES.json")),
    PRISON_RIP(ZombiesMap.PRISON, Difficulty.RIP, readFromFile("data/rounds/PRR_WAVES.json"));

    fun appliedDifficulty(difficulty: Difficulty) = when (map) {
        ZombiesMap.DEAD_END -> when (difficulty) {
            Difficulty.NORMAL -> DEAD_END_NORMAL
            Difficulty.HARD -> DEAD_END_HARD
            Difficulty.RIP -> DEAD_END_RIP
        }
        ZombiesMap.BAD_BLOOD -> when (difficulty) {
            Difficulty.NORMAL -> BAD_BLOOD_NORMAL
            Difficulty.HARD -> BAD_BLOOD_HARD
            Difficulty.RIP -> BAD_BLOOD_RIP
        }
        ZombiesMap.ALIEN_ARCADIUM -> ALIEN_ARCADIUM
        ZombiesMap.PRISON -> when (difficulty) {
            Difficulty.NORMAL -> PRISON_NORMAL
            Difficulty.HARD -> PRISON_HARD
            Difficulty.RIP -> PRISON_RIP
        }
    }

    override fun toString() = when (this) {
        DEAD_END_NORMAL -> "Dead End Normal"
        DEAD_END_HARD -> "Dead End Hard"
        DEAD_END_RIP -> "Dead End RIP"
        BAD_BLOOD_NORMAL -> "Bad Blood Normal"
        BAD_BLOOD_HARD -> "Bad Blood Hard"
        BAD_BLOOD_RIP -> "Bad Blood RIP"
        ALIEN_ARCADIUM -> "Alien Arcadium"
        PRISON_NORMAL -> "Prison Normal"
        PRISON_HARD -> "Prison Hard"
        PRISON_RIP -> "Prison RIP"
    }

    fun toStringForRecorder() = when (this) {
        DEAD_END_NORMAL -> "DEAD_END"
        DEAD_END_HARD -> "DEAD_END_HARD"
        DEAD_END_RIP -> "DEAD_END_RIP"
        BAD_BLOOD_NORMAL -> "BAD_BLOOD"
        BAD_BLOOD_HARD -> "BAD_BLOOD_HARD"
        BAD_BLOOD_RIP -> "BAD_BLOOD_RIP"
        ALIEN_ARCADIUM -> "ALIEN_ARCADIUM"
        PRISON_NORMAL -> "PRISON"
        PRISON_HARD -> "PRISON_HARD"
        PRISON_RIP -> "PRISON_RIP"
    }
}

private fun readFromFile(resourcePath: String): Array<Round> =
    JsonLoader.loadJsonFromResource(resourcePath, Array<Round>::class.java)